packages/analytics-react-native/android/src/main/java/com/amplitude/reactnative/LegacyDatabaseStorage.kt

```diff
@@ -0,0 +1,313 @@
+package com.amplitude.reactnative
+
+import android.content.Context
+import android.database.Cursor
+import android.database.sqlite.SQLiteDatabase
+import android.database.sqlite.SQLiteException
+import android.database.sqlite.SQLiteOpenHelper
+import org.json.JSONObject
+import java.io.File
+import java.util.LinkedList
+import java.util.Locale
+
+/**
+ * Store the database related constants.
+ * Align with com/amplitude/api/DatabaseHelper.java in previous SDK.
+ */
+object DatabaseConstants {
+    const val DEFAULT_INSTANCE = "\$default_instance"
+    const val DATABASE_NAME = "com.amplitude.api"
+    const val DATABASE_VERSION = 4
+
+    const val EVENT_TABLE_NAME = "events"
+    const val IDENTIFY_TABLE_NAME = "identifys"
+    const val IDENTIFY_INTERCEPTOR_TABLE_NAME = "identify_interceptor"
+    const val ID_FIELD = "id"
+    const val EVENT_FIELD = "event"
+
+    const val LONG_STORE_TABLE_NAME = "long_store"
+    const val STORE_TABLE_NAME = "store"
+    const val KEY_FIELD = "key"
+    const val VALUE_FIELD = "value"
+}
+
+/**
+ * The SDK doesn't need to write/read from local sqlite database.
+ * This storage class is used for migrating events only.
+ */
+class LegacyDatabaseStorage(context: Context, databaseName: String) : SQLiteOpenHelper(
+    context,
+    databaseName,
+    null,
+    DatabaseConstants.DATABASE_VERSION
+) {
+    private var file: File = context.getDatabasePath(databaseName)
+    var currentDbVersion: Int = DatabaseConstants.DATABASE_VERSION
+        private set
+
+    override fun onCreate(db: SQLiteDatabase) {
+        throw NotImplementedError()
+    }
+
+    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
+        currentDbVersion = oldVersion
+    }
+
+    private fun queryDb(
+        db: SQLiteDatabase,
+        table: String?,
+        columns: Array<String?>?,
+        selection: String?,
+        selectionArgs: Array<String?>?,
+        orderBy: String?,
+    ): Cursor? {
+        return db.query(table, columns, selection, selectionArgs, null, null, orderBy, null)
+    }
+
+    private fun handleIfCursorRowTooLargeException(e: java.lang.IllegalStateException) {
+        val message = e.message
+        if (!message.isNullOrEmpty() && message.contains("Couldn't read") && message.contains("CursorWindow")) {
+            closeDb()
+        } else {
+            throw e
```

Comment:
Will another migration happen if the first migration failed?

Reply:
@justin-fiedler  @yuhao900914 
I believe all exceptions are logged later, on JS side (needs some testing):
https://github.com/amplitude/Amplitude-TypeScript/blob/4ea90ed2640f822fb203b57a0745716834d9c2b6/packages/analytics-react-native/src/migration/remnant-data-migration.ts#L63-L78

> Will another migration happen if the first migration failed?

On next app startups another migrations happens (if config.migrateLegacy !== false) but identifies and intercepted identifies won't be migrated.

Reply:
Currently LegacyDatabaseStorage is a copy/paste of Amplitude-Android/Amplitude-iOS code. For simplicity I can remove all unnecessary exception handling and use `closeDB` for all exceptions

Reply:
Thanks, I think it makes sense to keep the code a direct copy of the legacy code.

Can you please test/confirm that any "native errors" are correctly handled via TS try/catch? 🙏

Reply:
@justin-fiedler 

> Can you please test/confirm that any "native errors" are correctly handled via TS try/catch?

I tested it and it doesn't work - the app crashes on "native" exceptions.
I added try/catch/log to Android native code.
I didn't add do/catch/log to iOS because current ios code can't throw exceptions (functions are not marked as `throws`):
![Screenshot_20230730_151205](https://github.com/amplitude/Amplitude-TypeScript/assets/9509929/6bb58e45-1d8b-4929-b2e5-d87014f41cce)